src/target/linux/readmem.rs

```diff
@@ -0,0 +1,287 @@
+        ReadOp::split_on_page_boundary(
+            &ReadOp {
+                remote_base,
+                len: mem::size_of::<T>(),
+                local_ptr: ptr as *mut _,
+            },
+            &mut self.read_ops,
+        );
+        self
+    }
+
+    /// Reads a slice of type `&mut [T]` from debuggee's memory at location `remote_base`.
+    /// This value will be written to the provided slice `val`.
+    /// You should call `apply` in order to execute the memory read operation.
+    /// The provided value `val` can't be accessed until either `apply` is called or `self` is
+    /// dropped.
+    ///
+    /// # Safety
+    ///
+    /// The type `T` must not have any invalid values.
+    /// For example, `T` must not be a `bool`, as `transmute::<u8, bool>(2)` is not a valid value for a bool.
+    /// In case of doubt, wrap the type in [`mem::MaybeUninit`].
+    // todo: further document mem safety - e.g., what happens in the case of partial read
+    pub unsafe fn read_slice<T>(mut self, val: &'a mut [T], remote_base: usize) -> Self {
+        ReadOp::split_on_page_boundary(
+            &ReadOp {
+                remote_base,
+                len: val.len() * mem::size_of::<T>(),
+                local_ptr: val.as_mut_ptr() as *mut _,
+            },
+            &mut self.read_ops,
+        );
+        self
+    }
+
+    /// Reads a `u8` byte slice from debuggee's memory at location `remote_base`.
+    /// This value will be written to the provided slice `val`.
+    /// You should call `apply` in order to execute the memory read operation.
+    pub fn read_byte_slice<T>(mut self, val: &'a mut [u8], remote_base: usize) -> Self {
+        ReadOp::split_on_page_boundary(
+            &ReadOp {
+                remote_base,
+                len: val.len(),
+                local_ptr: val.as_mut_ptr() as *mut _,
+            },
+            &mut self.read_ops,
+        );
+        self
+    }
+
+    /// Executes the memory read operation.
+    pub fn apply(self) -> Result<(), Box<dyn std::error::Error>> {
+        let read_len = self
+            .read_ops
+            .iter()
+            .fold(0, |sum, read_op| sum + read_op.len);
+
+        if read_len > isize::MAX as usize {
+            panic!("Read size too big");
+        };
+
+        // FIXME: Probably a better way to do this
+        let result = self.read_process_vm(&self.read_ops);
+
+        if result.is_err() && result.unwrap_err() == nix::Error::Sys(nix::errno::Errno::EFAULT)
+            || result.is_ok() && result.unwrap() != read_len as isize
+        {
+            let protected_maps = self
+                .target
+                .memory_maps()?
+                .into_iter()
+                .filter(|map| !map.is_readable)
+                .collect::<Vec<_>>();
+
+            let (protected, readable) =
+                split_protected(&protected_maps, self.read_ops.iter().cloned())?;
```

Comment:
You can use `self.read_ops.into_iter()` because `self` is passed by-value.

Reply:
> for maps to be an iterator.

Hmm, I guess then we'd need to convert it into a slice anyway, as `binary_search_by` is a `std::slice` method.
Or did you mean something else?

Reply:
:facepalm: